### PR TITLE
Secret management functions to DataUnion.ts in client

### PR DESCRIPTION
**Pull request by jtakalai**
Left to do:
* ~~add join-secret create/delete tests to default-join-server~~
* ~~add joining tests to default-join-server~~

Moved these into a new ticket ETH-344

### CODE REVIEW

File: packages/client/README.md [hunk 1]
```diff
@@ -88,6 +88,33 @@ Setting a new admin fee:
...
 const receipt = await dataUnion.setAdminFee(0.4)
 ```
+
+If the Data Union is set up to use the [default join server](../default-join-server/README.md) then members can join the Data Union by giving a correct secret.
+
+Admin can add secrets that allow anyone to join, as well as revoke those secrets, using the following functions:
+```js
+await dataUnion.createSecret() // returns the newly created secret
+await dataUnion.createSecret('user XYZ') // admin can also give the secret a more human-readable name
+await dataUnion.deleteSecret(secretId) // id is what was returned from createSecret
```
Comment (tumppi) [inline packages/client/README.md:98]: @jtakalai: `createSecret` does not return any id property. We should have one for `deleteSecret`.

**Comment by tumppi** [comment]
Old secret management also had support for editing secret's name. Could we have it here too?

File: packages/client/README.md (hunk 1 above)
Comment (jtakalai) [reply packages/client/README.md:98]: it's called `secret`, that property is the "id". Maybe it should be called id... but then it's also the actual thing that is passed. A bit confusing maybe.

**Comment by jtakalai** [comment]
> Old secret management also had support for editing secret's name. Could we have it here too?
I think let's make another ticket for making the API match old API. That would be a good thing for sure.

File: packages/client/README.md [hunk 2]
```diff
@@ -88,6 +88,33 @@ Setting a new admin fee:
 // Any number between 0 and 1, inclusive
 const receipt = await dataUnion.setAdminFee(0.4)
 ```
+
+If the Data Union is set up to use the [default join server](../default-join-server/README.md) then members can join the Data Union by giving a correct secret.
+
+Admin can add secrets that allow anyone to join, as well as revoke those secrets, using the following functions:
+```js
+await dataUnion.createSecret() // returns the newly created secret
```
Comment (kare) [inline packages/client/README.md:96]: Add return types to these examples and the ones below:
like: `const s: string = await dataUnion.createSecret()`

File: packages/client/package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "@dataunions/client",
-    "version": "3.0.0-beta.5",
+    "version": "3.0.0-beta.8",
```
Comment (kare) [inline packages/client/package.json:3]: `package-lock.json` requires an update if `package.json` is updated. `npm install -w packages/client` should do the trick.

File: packages/client/src/DataUnion.ts [hunk 4]
```diff
@@ -116,26 +123,140 @@ export class DataUnion {
...
             if (err.message?.match(/cannot estimate gas/)) {
-                throw new Error("Data Union join server couldn't send the join transaction. Please contact the join-server administrator.")
+                throw new Error("Data Union join-server couldn't send the join transaction. Please contact the join-server administrator.")
             }
             throw err
         })
     }
 
+    /**
+     * TODO: drop old DU support already probably...
```
Comment (kare) [inline packages/client/src/DataUnion.ts:165]: Better move TODO comments somewhere else from API documentation.

File: packages/client/src/DataUnion.ts [hunk 5]
```diff
@@ -116,26 +123,140 @@ export class DataUnion {
...
+            ]] = defaultAbiCoder.decode(['uint256[6]'], getStatsResponse) as BigNumber[][]
+            return {
+                totalEarnings,
+                totalWithdrawable: totalEarnings.sub(totalEarningsWithdrawn),
+                activeMemberCount,
+                inactiveMemberCount,
+                joinPartAgentCount,
+                lifetimeMemberEarnings,
+            }
+        } // TODO: maybe catch and re-throw with a better error message
```
Comment (kare) [inline packages/client/src/DataUnion.ts:210]: This TODO should be completed.

File: packages/default-join-server/.github/workflows/ci.yaml [hunk 6]
```diff
@@ -1,3 +1,4 @@
+# NOTE: this file is from the separate repo era and is NOT used (so probably outdated). TODO: delete?
```
Comment (kare) [inline packages/default-join-server/.github/workflows/ci.yaml:1]: This file should be deleted once the functionality is integrated to main CI build.

File: packages/client/test/unit/join.test.ts [hunk 7]
```diff
@@ -10,6 +10,11 @@ import type { DataUnionClientConfig } from '../../src/Config'
 
 import { deployContracts, getWallets } from './setup'
 
+/**
```
Comment (kare) [inline packages/client/test/unit/join.test.ts:17]: Note that this test is disabled in the master branch.

**Review by kare — COMMENTED** [review]
GitHub Actions comments should be easy to fix:
- Argument 'body' should be typed with a non-any type
- Missing return type on function

**Comment by jtakalai** [comment]
Tests actually passed, only Jest is segfaulting  >____<

File: packages/client/README.md (hunk 2 above)
Comment (jtakalai) [reply packages/client/README.md:96]: the return type is described in the next paragraph... I guess I could format it as TypeScript (or add TS examples) as well, but I think I'd prefer to make a separate ticket about that

File: packages/default-join-server/.github/workflows/ci.yaml (hunk 6 above)
Comment (jtakalai) [reply packages/default-join-server/.github/workflows/ci.yaml:1]: I wasn't sure to what extent it was... but sure, I'll delete it.

**Pull request merged**